# 

Normalize fields in VCF files generated by Varscan.
Specifically, AD field is rewritten as AD,RD and RD field is removed.


Documentation: http://varscan.sourceforge.net/somatic-calling.html
    AD - Reads supporting variant in tumor (or normal)
    RD - Reads supporting reference in tumor (or normal)

We want to change the value of AD to match that in mutect_vcf.  Specifically, we want,
    AD_new = RD_old,AD_old
    RD_new - delete it

Update meta-information
```
    ##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">
```

Delete ID=RD FORMAT field


For development and testing purposes, this project ships with test data and
demonstration scripts for running directly (in docker container), by calling
docker image, and by calling CWL tool. See ./testing for details.

Based on [MNP Filter CWL Tool](https://github.com/ding-lab/mnp_filter)

Preliminary analysis and discussion here: /gscuser/mwyczalk/projects/TinDaisy/CromwellRunner/adhoc-test/20191226.heiman/README.md

Docker image: 

Contact: Matt Wyczalkowski (m.wyczalkowski@wustl.edu)
